networking_aci/extensions/acioperations.py

```diff
@@ -0,0 +1,219 @@
+        hostgroup_name = kwargs.pop('id')
+        new_mode = kwargs.get("body", {}).get("mode")
+        if new_mode not in (aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL):
+            raise web_exc.HTTPBadRequest("Specify a mode that is either {} or {} (given mode was '{}')"
+                                         .format(aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL, new_mode))
+
+        ctx = context.get_admin_context()
+        curr_mode = self._get_mode(ctx, hostgroup_name)
+        if curr_mode == new_mode:
+            raise web_exc.HTTPBadRequest("{} is already in {} mode".format(hostgroup_name, new_mode))
+
+        # check if there are still  on this hostgroup
+        hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+        if not hg_config:
+            raise web_exc.HTTPBadRequest("Hostgroup {} has no config associated with it".format(hostgroup_name))
+
+        for segment_id in self.db.get_segment_ids_by_physnet(ctx, hg_config['physical_network']):
+            hosts = self.db.get_hosts_on_segment(ctx, segment_id)
+            for host in hg_config['hosts']:
+                if host in hosts:
+                    LOG.error("Denied API request to switch hostgroup %s to %s mode, host %s still bound in segment %s",
+                              hostgroup_name, new_mode, host, segment_id)
+                    raise web_exc.HTTPConflict("Could not switch, there is still a portbinding for host {} "
+                                               "present in segment {}"
+                                               .format(host, segment_id))
+
+        # cleanup old baremetal objects when switching away from baremetal mode
+        baremetal_bindings_cleaned = False
+        if hg_config['hostgroup_mode'] == aci_const.MODE_BAREMETAL:
+            LOG.info("Cleaning baremetal objects for %s from device", hostgroup_name)
+            try:
+                self.rpc_notifier.clean_baremetal_objects(hg_config)
+                baremetal_bindings_cleaned = True
+            except Exception:
+                LOG.exception("Could not clean baremetal bindings from ACI, cleaning will have to be done manually")
+
+        if self.db.set_hostgroup_mode(ctx, hostgroup_name, new_mode):
+            LOG.info("Hostgroup %s set to mode %s", hostgroup_name, new_mode)
+
+            hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+            aci_objects_update_succeeded = False
+            try:
+                # switching policy group of port selectors, etc.
+                self.rpc_notifier.sync_direct_mode_config(hg_config)
+                aci_objects_update_succeeded = True
+            except Exception:
+                LOG.exception("Could not update bindings on ACI, rpc call failed")
+
+            return {
+                "success": True,
+                "baremetal_bindings_cleaned": baremetal_bindings_cleaned,
+                "aci_objects_update_succeeded": aci_objects_update_succeeded,
```

Comment:
Just for my understand: How important are those two? If one of them is `False`, does that mean "successful, but you cannot work with that until checked manually"? If yes, how would the next automation request identify that this is the case?
"We need some follow-ups to make this work with automation. This is for manual use for now." is a valid answer here, too.

Reply:
In case this breaks we probably need manual intervention, yes. It should be usable by automation, but if it breaks we need to check what's what.

Reply:
I fear that people will just retry and neither Neutron nor ansible having state between jobs will think everything is fine ... :/

Maybe instead of bailing out with an error, if someone sets the same mode the hostgroup currently has, we could try to check if cleanup is still necessary? Basically making this command idempotent. It always returns success, if the wanted state is achieved.

Reply:
To clarify: The configuration is done again when the first infrastructure or baremetal machine is deployed in the hostgroup that is switched. It will then be done by the syncloop in periodic intervals als well. The only way to break this is by switching from baremetal to infra, then it needs to break on the ACI and then you deploy the machine by hand (ie without using ironic). So I think we are fine here for now. I think we'll approach this if this use-case arises.